packages/runtime/src/next/components/page.tsx

```diff
@@ -1,51 +1,82 @@
+  const client = useMakeswiftHostApiClient()
+
+  if (isServer) {
+    client.makeswiftApiClient.dispatch(updateAPIClientCache(cacheData))
+  }
+
+  useEffect(() => {
+    client.makeswiftApiClient.dispatch(updateAPIClientCache(cacheData))
+  }, [cacheData, client])
+}
+
+function useRegisterDocuments(rootDocuments: ReactPage.Document[]) {
+  const store = useStore()
+
+  return useEffect(() => {
+    const unregisterDocuments = rootDocuments.map(document =>
+      store.dispatch(registerDocumentEffect(document)),
+    )
+
+    return () => {
+      unregisterDocuments.forEach(unregisterDocument => {
+        unregisterDocument()
+      })
+    }
+  }, [store, rootDocuments])
+}
+
+function useRegisterPageDocument(pageDocument: MakeswiftPageDocument): ReactPage.Document {
+  const { locale, localizedPages, id, data } = pageDocument
+  const localizedPage = localizedPages.find(({ parentId }) => parentId == null)
+  const rootDocument = localizedPage
+    ? { key: localizedPage.elementTreeId, rootElement: localizedPage.data, locale }
+    : { key: id, rootElement: data, locale }
+
+  useRegisterDocuments([rootDocument])
+  return rootDocument
+}
+
 export const Page = memo(({ snapshot, ...props }: PageProps) => {
   if ('runtime' in props) {
     throw new Error(
       `The \`runtime\` prop is no longer supported in the \`@makeswift/runtime\` \`Page\` component as of \`0.15.0\`.
 See our docs for more information on what's changed and instructions to migrate: https://docs.makeswift.com/migrations/0.15.0`,
     )
   }
-  const client = useMemo(
-    () =>
-      new MakeswiftHostApiClient({
-        uri: new URL('graphql', snapshot.apiOrigin).href,
-        cacheData: snapshot.cacheData,
-        localizedResourcesMap: snapshot.localizedResourcesMap,
-        locale: snapshot.locale ? new Intl.Locale(snapshot.locale) : undefined,
-      }),
-    [snapshot],
-  )
 
-  const localizedPage = snapshot.document.localizedPages.find(({ parentId }) => parentId == null)
-  const rootElement = localizedPage
-    ? { key: localizedPage.elementTreeId, data: localizedPage.data }
-    : { key: snapshot.document.id, data: snapshot.document.data }
+  useCacheData(snapshot.cacheData)
 
-  const rootElements = new Map([[rootElement.key, rootElement.data]])
+  const rootDocument = useRegisterPageDocument(snapshot.document)
 
   return (
     <Suspense>
-      <RuntimeProvider client={client} rootElements={rootElements} preview={snapshot.preview}>
-        {/* We use a key here to reset the Snippets state in the PageMeta component */}
-        <PageMeta
-          key={snapshot.document.data.key}
-          pageDocument={snapshot.document}
-          documentKey={rootElement.key}
-        />
-      </RuntimeProvider>
+      {/* We use a key here to reset the Snippets state in the PageMeta component */}
+      <PageComponent
+        key={snapshot.document.data.key}
+        page={snapshot.document}
+        rootDocument={rootDocument}
+      />
```

Comment:
Pass the whole `rootDocument` instead of just the `documentKey`, see my earlier comment.